A36-xds-for-servers.md

```diff
@@ -0,0 +1,401 @@
+[A29]: A29-xds-tls-security.md
+
+## Proposal
+
+Each language will create an "XdsServer" API that will wrap the normal "Server"
+API. This will allow the implementation to see the listening ports, start event,
+stop event, and also allow it access to the server to change behavior (e.g., via
+an interceptor). Users will use the XdsServer API to construct the server
+instead of the existing API. XdsServer will not support ephemeral ports as part
+of this gRFC but may be enhanced in the future.
+
+Credential configuration will be managed separately by providing
+XdsServerCredentials, similar to client-side. The user must pass the
+XdsServerCredentials to the XdsServer API to enable control-plane managed TLS
+certificates. While the user-facing API is discussed here, the precise
+implementation details are covered in gRFC A29. The user is free to pass other
+ServerCredentials types to XdsServer and they will be honored. Note that RBAC is
+authz, not authn, so it would be enabled by using the XdsServer, even without
+XdsServerCredentials.
+
+### Serving and Not Serving
+
+To serve RPCs, the XdsServer must have its xDS configuration, provided via a
+Listener resource and potentialy other related resources. When its xDS
+configuration is missing or unknown the server must be in a "not serving" mode.
+This is ideally implemented by not `listen()`ing on the port. If that is
+impractical an implementation may be `listen()`ing on the port, but it must also
+`accept()` and immediately `close()` connections, making sure to not send any
+data to the client (e.g., no TLS ServerHello nor HTTP/2 SETTINGS). With either
+behavior, client connection attempts will quickly fail and clients would not
+perform further attempts without a backoff. Load balancing policies like
+`pick_first` would naturally attempt connections to any remaining addresses to
+quickly find an operational backend. However, the `accept()`+`listen()` approach
+will be improperly detected as server liveness for TCP heath checking.
+
+If the xDS bootstrap is missing or invalid, implementations would ideally fail
+XdsServer startup, but it is also acceptable to consider it a lack of xDS
+configuration and enter a permanent "not serving" mode.
+
+If the server is unable to open serving ports (e.g., because the port is already
+in use by another process), XdsServer startup may fail or it may enter "not
+serving" mode. If entering "not serving" mode, opening the port must be retried
+automatically (e.g., retry every minute).
+
+Communication failures do not impact the XdsServer's xDS configuration; the
+XdsServer should continue using the most recent configuration until connectivity
+is restored. However, XdsServer must accept configuration changes provided by
+the xDS server, including resource deletions. If that causes the XdsServer to
+lack essential configuration (e.g., the Listener was deleted) the server would
+need to enter "not serving" mode. "Not serving" requires existing connections to
```

Comment:
How about the "Not serving" to "Serving" transition? Specifically:

- this transition will automatically start processing RPCs without any intervention by the application, correct? i.e. it is as though `start()` was called on the XdsServer object (but without the application having to call it)
- this will require recreating the internal/delegate Server object since the delegate does not support restarting after shutting down. Is that correct?

It will be good to add some clarification for this.

Reply:
> this transition will automatically start processing RPCs without any intervention by the application, correct? i.e. it is as though start() was called on the XdsServer object (but without the application having to call it)

Yes. The application had already called `start()` as is expecting the server to be running. It was the library that "called stop()" so it is its responsibility to resume and restore the expected application state when able.

> this will require recreating the internal/delegate Server object since the delegate does not support restarting after shutting down. Is that correct?

Correct. This will vary with the language-specific API and the approach taken to become "not serving." But assuming XdsServer calls `server.stop()` on the wrapped server (the internal one), then it'd need to create a new one to replace it when calling `start()`.

I didn't go into more detail _here_ because that is language and implementation-specific. I did mention in the Java section that we may need to rebuild the server, and that was to allow for things here. But it isn't required to re-create the server; I left open the option for you to use a ProtocolNegotiator to immediately close() connections. Such an approach wouldn't have needed to re-create the server.